report/report.go

```diff
@@ -341,6 +341,18 @@ func (r Report) Validate() error {
 //
 // This for now creates node's LatestControls from Controls.
 func (r Report) Upgrade() Report {
+	needUpgrade := false
+	r.WalkTopologies(func(topology *Topology) {
+		for _, node := range topology.Nodes {
+			if node.LatestControls.Size() == 0 && len(node.Controls.Controls) > 0 {
+				needUpgrade = true
```

Comment:
An early exit would be nice here, but that would require introducing an alternative to `WalkTopologies`, which just isn't worth the effort, given that we hit this branch when an upgrade _is_ required, and upgrading is orders of magnitudes more expensive than doing this walk to the end, so we'd be saving relatively little.